Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1100 @@
+    ///
+    /// The `success` case indicates the RPC was accepted by the server for
+    /// processing, however, the RPC may still fail by throwing an error from its
+    /// `messages` sequence. The `failure` case indicates that the RPC was
+    /// rejected by the server.
+    public var result: Result<Body, RPCError>
+
+    public init(result: Result<Body, RPCError>) {
+      // ...
+    }
+
+    // Note: it's possible to provide a number of conveneince APIs on top:
+
+    /// The metadata received from server at the start of the RPC.
+    ///
+    /// The metadata will be empty if `result` is `failure`.
+    public var metadata: Metadata {
+      get {
+        // ...
+      }
+    }
+
+    /// The stream of messages received from the server.
+    public var messages: any AsyncSequence<Message> {
+      get {
+        // ...
+      }
+    }
+
+    /// The metadata received from server at the end of the RPC.
+    public var trailingMetadata: Metadata {
+      get {
+        // ...
+      }
+    }
+  }
+}
+
+// MARK: - Supporting types
+
+/// A sink for values which are produced over time.
+public protocol Writer<Element: Sendable>: Sendable {
+  /// Write a sequence of elements.
+  ///
+  /// Writes may suspend if the sink is unable to accept writes.
+  func write(contentsOf elements: some Sequence<Element>) async throws
+}
+
+extension Writer {
+  /// Write a single element.
+  public func write(_ element: Element) async throws {
+    try await self.write(contentsOf: CollectionOfOne(element))
+  }
+}
+
+/// An RPC error.
+///
+/// Every RPC is terminated with a status which includes a code, and optionally,
+/// a message. The status describes the ultimate outcome of the RPC.
+///
+/// This type is like a status but only represents negative outcomes, that is,
+/// all status codes except for `ok`. This type can also carry ``Metadata``.
+/// This can be used by service authors to transmit additional information to
+/// clients if an RPC throws an error.
+public struct RPCError: Error, Hashable, Sendable {
+  public struct Code: Hashable, Sendable {
+    public var code: UInt8
+
+    private init(_ code: UInt8) {
+      // ...
+    }
+
+    // All non-zero status codes from:
+    // https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+
+    /// The operation was cancelled (typically by the caller).
+    public static let cancelled = Self(code: 1)
+
+    /// Unknown error. An example of where this error may be returned is if a
+    /// status value received from another address space belongs to an error-space
+    /// that is not known in this address space. Also errors raised by APIs that
+    /// do not return enough error information may be converted to this error.
+    public static let unknown = Self(code: 2)
+
+    // etc.
+  }
+
+  /// The error code.
+  public var code: Code
+
+  /// A message describing the error.
+  public var message: String
+
+  /// Metadata associated with the error.
+  public var metadata: Metadata
```

Comment:
Interesting. Are we expecting to put the detailed error trailer in here or just all trailers>

Reply:
All trailers. Initially I didn't have `metadata` in the error type at all and ended up with a wrapper struct pairing an error with metadata which also conformed to `Error`. It was quite unpleasant to work with (do I catch `RPCError` or `RPCErrorWithMetadata`? Or both?).

Ultimately we expect this to represent the final state of a failed RPC where we'd expect to have metadata as well. That's not always the case because the final error can be synthesised on the client side but using an empty `Metadata` in that case is simpler than having to deal with two different error types anyway.

Reply:
I agree. This seems like a nice API, I was just confused by the document because it says `associated with the error` and I wasn't sure what that meant.